File: backend/src/main/java/com/ecommerce/service/ProductService.java
package com.ecommerce.service;

import com.ecommerce.model.Product;
import com.ecommerce.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    public Page<Product> getAllProducts(Pageable pageable) {
        return productRepository.findAll(pageable);
    }

    public Optional<Product> getProductById(Long id) {
        return productRepository.findById(id);
    }

    public List<Product> getProductsByCategory(String category) {
        return productRepository.findByCategory(category);
    }

    public List<Product> getProductsByBrand(String brand) {
        return productRepository.findByBrand(brand);
    }

    public Page<Product> searchProducts(String keyword, Pageable pageable) {
        return productRepository.findByKeyword(keyword, pageable);
    }

    public List<String> getAllCategories() {
        return productRepository.findAllCategories();
    }

    public List<String> getAllBrands() {
        return productRepository.findAllBrands();
    }

    public List<Product> getAvailableProducts() {
        return productRepository.findByStockQuantityGreaterThan(0);
    }

    public Product saveProduct(Product product) {
        return productRepository.save(product);
    }

    public Product updateProduct(Long id, Product productDetails) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + id));

        product.setName(productDetails.getName());
        product.setDescription(productDetails.getDescription());
        product.setPrice(productDetails.getPrice());
        product.setStockQuantity(productDetails.getStockQuantity());
        product.setImageUrl(productDetails.getImageUrl());
        product.setCategory(productDetails.getCategory());
        product.setBrand(productDetails.getBrand());

        return productRepository.save(product);
    }

    public void deleteProduct(Long id) {
        productRepository.deleteById(id);
    }

    public boolean isProductAvailable(Long productId, Integer quantity) {
        Optional<Product> product = productRepository.findById(productId);
        return product.isPresent() && product.get().getStockQuantity() >= quantity;
    }

    public void updateStock(Long productId, Integer quantity) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with id: " + productId));
        
        int newStock = product.getStockQuantity() - quantity;
        if (newStock < 0) {
            throw new RuntimeException("Insufficient stock for product: " + product.getName());
        }
        
        product.setStockQuantity(newStock);
        productRepository.save(product);
    }
} 

File: frontend/src/services/productService.js
import api from './api';

export const productService = {
  getAllProducts: (params = {}) => {
    const { page = 0, size = 12, sortBy = 'id', sortDir = 'asc' } = params;
    return api.get(`/products?page=${page}&size=${size}&sortBy=${sortBy}&sortDir=${sortDir}`)
      .then(response => response.data);
  },

  getProductById: (id) => {
    return api.get(`/products/${id}`).then(response => response.data);
  },

  searchProducts: (keyword, params = {}) => {
    const { page = 0, size = 12 } = params;
    return api.get(`/products/search?keyword=${keyword}&page=${page}&size=${size}`)
      .then(response => response.data);
  },

  getProductsByCategory: (category) => {
    return api.get(`/products/category/${category}`).then(response => response.data);
  },

  getProductsByBrand: (brand) => {
    return api.get(`/products/brand/${brand}`).then(response => response.data);
  },

  getCategories: () => {
    return api.get('/products/categories').then(response => response.data);
  },

  getBrands: () => {
    return api.get('/products/brands').then(response => response.data);
  },

  getAvailableProducts: () => {
    return api.get('/products/available').then(response => response.data);
  },

  createProduct: (productData) => {
    return api.post('/products', productData).then(response => response.data);
  },

  updateProduct: (id, productData) => {
    return api.put(`/products/${id}`, productData).then(response => response.data);
  },

  deleteProduct: (id) => {
    return api.delete(`/products/${id}`).then(response => response.data);
  }
}; 